cli/run_util.go

```diff
@@ -42,440 +42,460 @@ import (
+	results := r.activateAndRunSteps(plan, steplibSource, buildRunResults, environments, secrets, isLastWorkflow, tracker, workflowIDProperties)
+
+	tracker.SendWorkflowFinished(workflowIDProperties, results.IsBuildFailed())
+	collectToolVersions(tracker)
+
+	return results
+}
+
+func (r WorkflowRunner) activateAndRunSteps(
+	plan models.WorkflowExecutionPlan,
+	defaultStepLibSource string,
+	buildRunResults models.BuildRunResultsModel,
+	environments *[]envmanModels.EnvironmentItemModel,
+	secrets []envmanModels.EnvironmentItemModel,
+	isLastWorkflow bool,
+	tracker analytics.Tracker,
+	workflowIDProperties coreanalytics.Properties,
+) models.BuildRunResultsModel {
+	log.Debug("[BITRISE_CLI] - Activating and running steps")
+
+	if len(plan.Steps) == 0 {
+		log.Warnf("%s workflow has no steps to run, moving on to the next workflow...", plan.WorkflowTitle)
+		return buildRunResults
 	}
 
-	for _, env := range inventoryEnvironments {
-		key, value, err := env.GetKeyValuePair()
-		if err != nil {
-			return false, err
+	// ------------------------------------------
+	// In function global variables
+
+	// These are global for easy use in local register step run result methods.
+	runResultCollector := newBuildRunResultCollector(tracker)
+
+	// Global variables for synchronising container's lifecycle
+	// and shutting down the last set of containers at the end of the workflow run
+	currentStepGroupID := ""
+	lastStepGroupID := ""
+	defer func() {
+		if lastStepGroupID != "" {
+			r.stopContainersForStepGroup(lastStepGroupID, plan.WorkflowTitle)
 		}
+	}()
 
-		if key == configs.PullRequestIDEnvKey && value != "" {
-			return true, nil
+	// ------------------------------------------
+	// Main - Preparing & running the steps
+	for idx, stepPlan := range plan.Steps {
+		if stepPlan.GroupID != currentStepGroupID {
+			if currentStepGroupID != "" {
+				r.stopContainersForStepGroup(currentStepGroupID, plan.WorkflowTitle)
+			}
+
+			if stepPlan.GroupID != "" {
+				if len(stepPlan.ContainerID) > 0 || len(stepPlan.ServiceIDs) > 0 {
+					r.startContainersForStepGroup(stepPlan.ContainerID, stepPlan.ServiceIDs, *environments, stepPlan.GroupID, plan.WorkflowTitle)
+				}
+			}
+
+			currentStepGroupID = stepPlan.GroupID
+			if stepPlan.GroupID != "" {
+				lastStepGroupID = stepPlan.GroupID
+			}
 		}
-		if key == configs.PRModeEnvKey && value == "true" {
-			return true, nil
+
+		stepStartTime := time.Now()
+		stepIDProperties := coreanalytics.Properties{analytics.StepExecutionID: stepPlan.UUID}
+		stepStartedProperties := workflowIDProperties.Merge(stepIDProperties)
+
+		result := r.activateAndRunStep(
+			stepPlan.Step,
+			stepPlan.StepID,
+			idx,
+			defaultStepLibSource,
+			stepPlan.UUID,
+			tracker,
+			*environments,
+			secrets,
+			buildRunResults,
+			plan.IsSteplibOfflineMode,
+			stepPlan.ContainerID,
+			stepPlan.GroupID,
+			stepStartTime,
+			stepStartedProperties,
+		)
+
+		*environments = append(*environments, result.OutputEnvironments...)
```

Comment:
This is the single place where a step run modifies the "build"'s env vars, subsequent functions are just reading the env vars, and that's when they receive environments as a value argument.